packages/wrangler/src/api/startDevWorker/ProxyController.ts

```diff
@@ -0,0 +1,531 @@
+import path from "node:path";
+import { LogLevel, Miniflare, Response } from "miniflare";
+import { WebSocket } from "miniflare";
+import inspectorProxyWorkerPath from "worker:startDevWorker/InspectorProxyWorker";
+import proxyWorkerPath from "worker:startDevWorker/ProxyWorker";
+import {
+	logConsoleMessage,
+	maybeHandleNetworkLoadResource,
+} from "../../dev/inspect";
+import { WranglerLog, castLogLevel } from "../../dev/miniflare";
+import { getHttpsOptions } from "../../https-options";
+import { logger } from "../../logger";
+import { getSourceMappedStack } from "../../sourcemap";
+import { castErrorCause } from "./events";
+import { assertNever, createDeferred, type DeferredPromise } from "./utils";
+import type { EsbuildBundle } from "../../dev/use-esbuild";
+import type {
+	BundleStartEvent,
+	ConfigUpdateEvent,
+	ErrorEvent,
+	InspectorProxyWorkerIncomingWebSocketMessage,
+	InspectorProxyWorkerOutgoingRequestBody,
+	InspectorProxyWorkerOutgoingWebsocketMessage,
+	PreviewTokenExpiredEvent,
+	ProxyData,
+	ProxyWorkerIncomingRequestBody,
+	ProxyWorkerOutgoingRequestBody,
+	ReadyEvent,
+	ReloadCompleteEvent,
+	ReloadStartEvent,
+	SerializedError,
+} from "./events";
+import type { StartDevWorkerOptions } from "./types";
+import type { MiniflareOptions } from "miniflare";
+
+export class ProxyController extends EventEmitter {
+	public ready = createDeferred<ReadyEvent>();
+
+	public proxyWorker?: Miniflare;
+	public inspectorProxyWorker?: Miniflare;
+	proxyWorkerOptions?: MiniflareOptions;
+	inspectorProxyWorkerOptions?: MiniflareOptions;
+	inspectorProxyWorkerWebSocket?: DeferredPromise<WebSocket>;
+
+	protected latestConfig?: StartDevWorkerOptions;
+	protected latestBundle?: EsbuildBundle;
+	secret = randomUUID();
+
+	protected createProxyWorker() {
+		if (this._torndown) return;
+		assert(this.latestConfig !== undefined);
+
+		const cert =
+			this.latestConfig.dev?.server?.secure ||
+			this.latestConfig.dev?.inspector?.secure
+				? getHttpsOptions()
+				: undefined;
+
+		const proxyWorkerOptions: MiniflareOptions = {
+			verbose: true,
+			compatibilityFlags: ["nodejs_compat"],
+			modulesRoot: path.dirname(proxyWorkerPath),
+			modules: [{ type: "ESModule", path: proxyWorkerPath }],
+			durableObjects: {
+				DURABLE_OBJECT: "ProxyWorker",
+			},
+			serviceBindings: {
+				PROXY_CONTROLLER: async (req): Promise<Response> => {
+					const message = (await req.json()) as ProxyWorkerOutgoingRequestBody;
+
+					this.onProxyWorkerMessage(message);
+
+					return new Response(null, { status: 204 });
+				},
+			},
+			bindings: {
+				PROXY_CONTROLLER_AUTH_SECRET: this.secret,
+			},
+
+			// upsttream is how we tell miniflare what request.url should look like inside the worker
+			upstream:
+				this.latestConfig.dev?.urlOverrides &&
+				new URL(
+					`${this.latestConfig.dev.urlOverrides.secure ? "https" : "http"}://${
+						this.latestConfig.dev.urlOverrides.hostname
+					}`
+				).href,
```

Comment:
There was a time where the ProxyWorker used the request.url.hostname so I set upstream and set the proxy header 'MF-Original-URL' but now we don't need hostname in the ProxyWorker we can go back to not setting upstream and rely on the UserWorker miniflare instance's upstream option – this will save a ProxyWorker reload when this config changes too

Reply:
Changed it to use a new property on proxyData and not rely on Miniflare the `upstream` option anywhere